pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,308 @@
+	defer func() {
+		if err != nil {
+			return
+		}
+		if groupKey == "" {
+			c.metrics.Hits.Inc(1)
+		} else {
+			c.metrics.Misses.Inc(1)
+		}
+	}()
+	for {
+		cached, ok := c.getHydratedTableDescriptorFromCache(ctx, key)
+		if ok {
+			canUse, err := canUseCachedDescriptor(ctx, cached, res)
+			if err != nil {
+				return nil, err
+			}
+			if canUse {
+				return cached.tableDesc, nil
+			}
+		}
+		if groupKey == "" {
+			groupKey = fmt.Sprintf("%d@%d", key.ID, key.Version)
+		}
+		// Now we want to lock the cache and populate the descriptors.
+		// Only the calling goroutine can actually use the result directly.
+		// Furthermore, if an error is encountered, only the calling goroutine
+		// should return it.
+		var called bool
+		res, _, err := c.g.Do(groupKey, func() (interface{}, error) {
+			called = true
+			cachedRes := cachedTypeDescriptorResolver{
+				underlying: res,
+				cache:      map[descpb.ID]*cachedType{},
+			}
+			descBase := protoutil.Clone(table.TableDesc()).(*descpb.TableDescriptor)
+			if err := typedesc.HydrateTypesInTableDescriptor(ctx, descBase, &cachedRes); err != nil {
+				return nil, err
+			}
+			hydrated := tabledesc.NewImmutable(*descBase)
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			c.mu.cache.Add(key, &hydratedTableDescriptor{
+				tableDesc: hydrated,
+				typeDescs: cachedRes.types,
+			})
+			return hydrated, nil
+		})
+		if !called {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		return res.(*tabledesc.Immutable), nil
+	}
+}
+
+type cachedTypeDescriptorResolver struct {
+	underlying catalog.TypeDescriptorResolver
+	cache      map[descpb.ID]*cachedType
+	types      []*cachedType
+}
+
+func (c *cachedTypeDescriptorResolver) GetTypeDescriptor(
+	ctx context.Context, id descpb.ID,
+) (tree.TypeName, catalog.TypeDescriptor, error) {
+	if typ, exists := c.cache[id]; exists {
+		return typ.name, typ.typDesc, nil
+	}
+	name, typDesc, err := c.underlying.GetTypeDescriptor(ctx, id)
+	if err != nil {
+		return tree.TypeName{}, nil, err
+	}
+	typ := &cachedType{
+		name:    name,
+		typDesc: typDesc,
+	}
+	c.cache[id] = typ
+	c.types = append(c.types, typ)
+	return name, typDesc, nil
+}
+
+func canUseCachedDescriptor(
+	ctx context.Context, cached *hydratedTableDescriptor, res catalog.TypeDescriptorResolver,
+) (canUse bool, _ error) {
+	for _, typ := range cached.typeDescs {
+		name, typDesc, err := res.GetTypeDescriptor(ctx, typ.typDesc.GetID())
+		if err != nil {
+			return false, err
+		}
+		// Ensure that the type version matches.
+		if typ.typDesc.GetVersion() != typDesc.GetVersion() ||
+			// Only match on the name if the retrieved type has a qualified name.
+			// Note that this behavior is important and ensures that when this
+			// function is passed a resolved which looks up qualified names, it always
+			// get a hydrated descriptor with qualified names in its types.
+			// This is important as we share this cache between distsql which does
```

Comment:
I'm still confused. Right now, the cache isn't accessed by DistSQL (we only resolve types through the lease manager, unless you're planning for the future now). Based on the code above, if this returns false, we'll still go and put the descriptor into the cache with the same key, so we would thrash right?

Reply:
right, right there's a bit more to do here. I'm not sure I follow what you mean about thrashing. It shouldn't thrash back and forth. Sure a resolved name should kick out an unresolved name but not vice-versa. 

I've had two different approach I've been thinking about:

1) have distsql resolve the table using the machinery (except in the case where it's a modified table) but in that case I believe the distsql stuff is set up to not
2) Augment the cache built here to have a cache for the underlying `*types.T` values and utilize those when hydrating descriptors to avoid just a few allocs. 

1) is better but bigger is my sense. In either case, we want the above logic I think.